### ISSUE TITLE
firewall infrastructure does not get reconciled

### ISSUE THREAD

**Issue by mwennrich** [metal-stack/gardener-extension-provider-metal#15]
How to reproduce:

- create cluster
- cluster has two workers and one firewall:

```
metalctl machine list --project ed53d0dc-fd1c-41f9-a534-bf84c61c98c9
  ID                                         LAST EVENT   WHEN  AGE      HOSTNAME                        PROJECT                               SIZE           IMAGE         PARTITION
  fafd0c00-7090-11e9-8000-efbeaddeefbe       Phoned Home  2s    23m 9s   shoot--p7l8m...-firewall-fcbab  ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  c1-xlarge-x86  Firewall 1    nbg-w8101
  00000000-beef-beef-0011-efbeaddeefbe       Phoned Home  3s    18m 52s  shoot--p7l8m...dfb8b55dc-f9zxg  ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  c1-xlarge-x86  Ubuntu 19.04  nbg-w8101
  00000000-beef-beef-0001-efbeaddeefbe       Phoned Home  1s    6m 36s   shoot--p7l8m...dfb8b55dc-nfvv9  ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  c1-xlarge-x86  Ubuntu 19.04  nbg-w8101

```
- destroy firewall
`metalctl machine rm fafd0c00-7090-11e9-8000-efbeaddeefbe
`
- reconcile cluster
`cloudctl cluster reconcile c1605f39-1668-11ea-853c-42ced4c4a306`

=> reconcile freezes at 86%:

```
  UID                                   TENANT  PROJECT                               NAME      VERSION  PARTITION  OPERATION   PROGRESS         API   CONTROL  NODES  SYSTEM  SIZE  AGE
  c1605f39-1668-11ea-853c-42ced4c4a306  fits    ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  mwen4895  1.14.3   nbg-w8101  Processing  86% [Reconcile]  True  True     False  False   2/2   54m 7s

```
project still without firewall:
```
  ID                                         LAST EVENT   WHEN  AGE      HOSTNAME                        PROJECT                               SIZE           IMAGE         PARTITION
  00000000-beef-beef-0011-efbeaddeefbe       Phoned Home  15s   49m 6s   shoot--p7l8m...dfb8b55dc-f9zxg  ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  c1-xlarge-x86  Ubuntu 19.04  nbg-w8101
  00000000-beef-beef-0001-efbeaddeefbe       Phoned Home  17s   36m 50s  shoot--p7l8m...dfb8b55dc-nfvv9  ed53d0dc-fd1c-41f9-a534-bf84c61c98c9  c1-xlarge-x86  Ubuntu 19.04  nbg-w8101

```

**Comment by Gerrit91** [comment]
That's right. We do not ask the metal-api again after we got a machine id from the firewall create request. Would be a good thing to make this more robust.

**Comment by Gerrit91** [comment]
Unfortunately, the Gardener does not explicitly request infrastructure reconciliation once it succeeded successfully. When adding the `gardener.cloud/operation: "reconcile"` annotation to the infrastructure resource, the newly implemented behavior of recreating a firewall works fine...

**Comment by rfranzke** [comment]
@Gerrit91 The Gardener does add the `gardener.cloud/operation=reconcile` annotation to the `Infrastructure` only during the maintenance time window. It's a little bit special here. For all other extension resources (like `Worker`, `ControlPlane`, etc), it adds the annotation in *every* reconciliation loop. The rational was to minimize the number of calls to the IaaS providers to prevent running into quota limits, especially if you have a lot of shoots using the same IaaS account. Hence, the idea was to run the infrastructure reconciliation only once a day - in the maintenance time window. WDYT?

**Comment by rfranzke** [comment]
Btw, to trigger an immediate maintenance operation (if you don't want to wait till the time window starts) you can annotate your shoot with `shoot.garden.sapcloud.io/operation=maintain` instead of `shoot.garden.sapcloud.io/operation=reconcile`.

**Comment by Gerrit91** [comment]
Rafael, I did not expect you here! :D 

Thanks for clarification. We wanted to make our firewall creation idempotent. I guess it's fine for the time being if the infrastructure gets only reconciled during maintenance window. We differ a little from the other cloud providers here because we actually maintain a server in the infrastructure. Ideally, this would be managed by something like the `machine-controller-machine`, such that we can also do machine upgrades. Maybe we even need a special controller to handle this properly... not sure yet.

**Comment by Gerrit91** [comment]
We now support updating the firewall spec (image or size), which causes firewall deletion (hence downtime) and recreation.

**Comment by Gerrit91** [comment]
I think, we can close this for now as it has partially been resolved. Let's open another issue for "zero-downtime firewall updates" when it's time.
